src/components/tree-view/tree-view.tsx

```diff
@@ -25,37 +28,77 @@ export interface TreeViewProps {
     selectedItem?: TreeItemData;
 }
 
+export interface TreeItemState {
+    isSelected: boolean;
+}
+
+export type StateMap = Map<TreeItemData, TreeItemState>;
+
 const itemIdPrefix = 'TREE_ITEM';
 
-export function TreeItem({ item, itemRenderer, onItemClick, isSelected }: TreeItemProps): JSX.Element {
+export function TreeItem({ item, itemRenderer, onItemClick, stateMap, state }: TreeItemProps): JSX.Element {
```

Comment:
@JoMarton i prefer this flavour as it reflects the nature of a functional component, in my opinion.

@VKobeliatsky i extends with `React.Attributes` because i need to play a `key` when i create the elements.